core/src/main/scala/akka/kafka/internal/ProducerStage.scala

```diff
@@ -6,126 +6,299 @@
+        case _ if awaitingConf > 0 =>
+          suspendDemand()
+          scheduleOnce(commitSchedulerKey, messageDrainIntervalMs.milliseconds)
+        case _ =>
+          scheduleOnce(commitSchedulerKey, commitIntervalMs.milliseconds)
       }
     }
-    logic
+
+    override val onMessageAckCb: AsyncCallback[Message[K, V, P]] =
+      getAsyncCallback[Message[K, V, P]](_.passThrough match {
+        case o: ConsumerMessage.PartitionOffset => batchOffsets = batchOffsets.updated(o)
+        case _ =>
+      })
+
+    override def onCompletionSuccess(): Unit = {
+      log.debug("Committing final transaction before shutdown")
+      cancelTimer(commitSchedulerKey)
+      maybeCommitTransaction(beginNewTransaction = false)
+      super.onCompletionSuccess()
+    }
+
+    override def onCompletionFailure(ex: Throwable): Unit = {
+      log.debug("Aborting transaction due to stage failure")
+      abortTransaction()
+      super.onCompletionFailure(ex)
+    }
+
+    private def commitTransaction(batch: NonemptyTransactionBatch, beginNewTransaction: Boolean): Unit = {
+      val group = batch.group
+      val offsetMap = batch.offsetMap().asJava
+      log.debug(s"Committing transaction for consumer group '$group' with offsets: $offsetMap")
+      producer.sendOffsetsToTransaction(offsetMap, group)
+      producer.commitTransaction()
+      batchOffsets = TransactionBatch.empty
+      if (beginNewTransaction) {
+        beginTransaction()
+        resumeDemand()
+        scheduleOnce(commitSchedulerKey, commitIntervalMs.milliseconds)
+      }
+    }
+
+    private def initTransactions(): Unit = {
+      log.debug("Initializing transactions")
+      producer.initTransactions()
+    }
+
+    private def beginTransaction(): Unit = {
+      log.debug("Beginning new transaction")
+      producer.beginTransaction()
+    }
+
+    private def abortTransaction(): Unit = {
+      log.debug("Aborting transaction")
+      producer.abortTransaction()
+    }
+  }
+
+  trait ProducerCompletionState {
+    def onCompletionSuccess(): Unit
+    def onCompletionFailure(ex: Throwable): Unit
+  }
+
+  trait MessageCallback[K, V, P] {
+    def awaitingConfirmation: AtomicInteger
+    def onMessageAckCb: AsyncCallback[Message[K, V, P]]
+  }
+
+  object TransactionBatch {
+    def empty: TransactionBatch = new EmptyTransactionBatch()
+  }
+
+  trait TransactionBatch {
+    def updated(partitionOffset: PartitionOffset): TransactionBatch
+  }
+
+  class EmptyTransactionBatch extends TransactionBatch {
+    override def updated(partitionOffset: PartitionOffset): TransactionBatch = new NonemptyTransactionBatch(partitionOffset)
+  }
+
+  class NonemptyTransactionBatch(
+      head: PartitionOffset,
+      tail: Map[GroupTopicPartition, Long] = Map[GroupTopicPartition, Long]())
+    extends TransactionBatch {
+    private val offsets = tail + (head.key -> head.offset)
+
+    def group: String = offsets.keys.head.groupId
```

Comment:
why not only `head.key.groupId` ?

Reply:
:+1:

Reply:
thinking of it you could probably simplify the NonemptyTransactionBatch by directly storing the offsets and groupId :

```
class EmptyTransactionBatch extends TransactionBatch {
    override def updated(partitionOffset: PartitionOffset): TransactionBatch = new NonemptyTransactionBatch(partitionOffsetkey.groupId, Map(partitionOffset.key -> partitionOffset.offset))
}

 class NonemptyTransactionBatch(val group : String, offsets : Map[GroupTopicPartition, Long]) {
         def offsetMap(): Map[TopicPartition, OffsetAndMetadata] = offsets.map {
           case (gtp, offset) => new TopicPartition(gtp.topic, gtp.partition) -> new OffsetAndMetadata(offset + 1)
        }

override def updated(partitionOffset: PartitionOffset): TransactionBatch = {
     require(
          group == partitionOffset.key.groupId,
        s"Transaction batch must contain messages from exactly 1 consumer group. $group != ${partitionOffset.key.groupId}")
 
         val updatedOffsets = offsets + (partitionOffset.key -> partitionOffset.offset)

      new NonemptyTransactionBatch(group, updateOffsets)
    }
}


}

Reply:
Passing in the group name may be a good idea.  I want to give the `NonemptyTransactionBatch` the property of never being empty, so I was inspired by the [`NonEmptyList` implementation in cats](https://github.com/typelevel/cats/blob/master/core/src/main/scala/cats/data/NonEmptyList.scala) to provide the `head` element in the default constructor.